File: OnlyTheLonley/src/main/scala/scalaSingleton/Singleton.scala
package scalaSingleton

object Db {
  def expensiveOperationOne = {
    println ("performing expensive operation one")
    "expensive operation one"
  }
  def expensiveOperationTwo = {
    println ("performing expensive operation two")
    "expensive operation two"
  }
}

object Singleton {
  println("Singleton Init Started")
  val x = Db.expensiveOperationOne
  lazy val y = Db.expensiveOperationTwo
  println("Singleton Init Done")
}

File: fodder/src/main/java/javaSingleton/Singleton.java
package javaSingleton;

public class Singleton {
	private static Singleton instance;
	
	private String x, y;
	
	private Singleton() {
		System.out.println("Singleton Init Started");
		x = Db.getInstance().expensiveOperationOne();
		System.out.println("Singleton Init Done");
	}

	public synchronized static Singleton getInstance() {
		if (instance == null) {
			instance = new Singleton();
		}
		return instance;
	}

	public String getX() {
		return x;
	}

	public synchronized String getY() {
		if(y == null) {
			y = Db.getInstance().expensiveOperationTwo();
		}
		return y;
	}
}
